### PR TITLE
Add initial set of unit tests for `data_builder`

### LINKED ISSUE #280: Add pytest unit tests to `data_builder`
It would be great to have some unit tests for the methods in `data_builder` to help find edge cases in data formatting. I'm happy to start up a suite using pytest - please let me know if I should proceed and I'll start a PR.

### PR DESCRIPTION
**Pull request by ShaynaCummings**
There's still more to do here, but wanted to put up something mergeable 

- Add `pytest` and `pytest-cov` for unit testing
- Add unit tests for `title_to_name_date` with some comments/TODOs for cases to handle
- Add tests for `process_md_texts`

Closes #280

### CODE REVIEW

**Comment by ShaynaCummings** [comment]
Would love your thoughts on this @ubershmekel, @samgoldman or any others

File: tools/data_builder.py [hunk 1]
```diff
@@ -119,6 +122,8 @@ def parse_state(state, text):
 
     if entry and entry["links"]:
         yield entry
+    else:
+        print(f"Failed links parse: missing links for {entry}")
```
Comment (ubershmekel) [inline tools/data_builder.py:126]: This is not a failure mode, I think it's the norm. Perhaps a comment is missing there, that it's trying to make sure if we didn't find another city or title, then to submit the last entry we had.

The thing to print in that `else` case is "Finished processing this state" if you really wanted to print something

File: tools/tests/test_data_builder.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+def test_handle_missing_name(capsys):
+    title_missing_name = '| May 30th'
+    title_to_name_date(title_missing_name)
+
+    captured = capsys.readouterr()
+    assert "Failed name parse: missing name for" in str(captured) 
+
+
+@pytest.mark.skip(reason="failing test, need to handle this case")
```
Comment (ubershmekel) [inline tools/tests/test_data_builder.py:30]: Does this issue exist in the data? If it does - then I think you shouldn't skip it. Should you?

**Comment by ubershmekel** [comment]
Thank you for this!

I added a few comments and approved so you can push at your discretion.

File: tools/data_builder.py (hunk 1 above)
Comment (ShaynaCummings) [reply tools/data_builder.py:130]: If you run this against the existing data set, this will only print on entries that have no links (e.g. Kansas.md's `(no entries yet)`) - the `yield` on 124 above is actually the 'finished' state

File: tools/tests/test_data_builder.py (hunk 2 above)
Comment (ShaynaCummings) [reply tools/tests/test_data_builder.py:30]: This issue does not exist in the current data but I included this just in case we wanted to update the validation to guard against it. Happy to take a stab at that after merging this if it's helpful

**Comment by ShaynaCummings** [comment]
@ubershmekel I've added the github actions and the updated workflow with pytest appears to be working against my fork, which is great! 

One question I have for you is: currently these unit tests use fixture data. I could change the workflow to run the tests as both unit tests _and_ against the real data outputted from the `data_builder`. That way, the CI build would fail if any of the actual data in the compiled file failed validation as defined by the tests.

Also, if you wanted to proceed with that, a further question would be, which validation failures should specifically break CI versus just print a warning in the output?

File: tools/data_builder.py (hunk 1 above)
Comment (ubershmekel) [reply tools/data_builder.py:130]: Then I misunderstood. Thank you.

**Comment by ubershmekel** [comment]
@ShaynaCummings the tests (fixture and real data) should happen on pull requests for every change, not just on push like the data build does. I've added another `yml` for that.

With regards to what should fail the test - at this point I would say don't prevent a merge, but visibly fail tests. Once we confirm the tests are reliably identifying problems - let's then prevent merges.

Also - I'd rather we do small changes, rather than one big PR, so let's merge this and open more.
